merlin-sdk/src/main/java/gov/nasa/jpl/aerie/merlin/protocol/json/SerializedValueSerializer.java

```diff
@@ -0,0 +1,113 @@
+package gov.nasa.jpl.aerie.merlin.protocol.json;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.ser.std.StdSerializer;
+import gov.nasa.jpl.aerie.merlin.protocol.types.SerializedValue;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public final class SerializedValueSerializer extends StdSerializer<SerializedValue> {
+
+  public SerializedValueSerializer() {
+    super((Class<SerializedValue>) null);
+  }
+  public SerializedValueSerializer(Class<SerializedValue> c) {
+    super(c);
+  }
+
+  @Override
+  public void serialize(final SerializedValue value, final JsonGenerator gen, final SerializerProvider provider)
+  throws IOException
+  {
+    value.match(new SerializedValue.Visitor<>() {
+      @Override
+      public Object onNull() {
+        try {
+          gen.writeNull();
+        } catch (IOException e) {
+          e.printStackTrace();
```

Comment:
Might want to consider throwing an `Error` here.

Reply:
A couple other changes I'd make would be to:
- Explicitly make the visitor generic type `Void`. 
- Wrap the `IOException` handling.

For the `IOException` handling you could make some interface like:
```java
private interface JsonGeneratorRunnable {
  void run() throws IOException;
}
```
And make a method that takes one of those as an argument:
```java
private static void tryGenerate(final JsonGeneratorRunnable r) {
  try {
    r.run();
  } catch (final IOException e) {
    throw new Error(...);
  }
}
```
This is just a rough outline, you could change things around to also report the type that was trying to be generated.

Reply:
done

Reply:
Oops, sorry I didn't see your second comment here. I ended up returning the caught error outside of the visitor and throwing it there. This way it doesn't require changing the SerializedValue visitor interface to allow errors. Is that ok?

Reply:
Ohh I see! Was wondering why it was changed to return an `IOException` 😉

Sorry to be a pain here but I'm not sure how I feel about catching and returning the `IOException`. This means that callers can decide if they want to ignore the serialization error or not. Instead of giving callers the choice (which by the "default" path of least resistance they will ignore) I think I prefer always throwing an exception/error.

Currently within Aerie the convention has been to throw a `RuntimeException` (see https://stackoverflow.com/questions/20461688/runtimeexception-error for a short explanation for why `Error` is probably not preferred here). If you search for `getSuccessOrThrow()` methods you'll see that pattern.

So for convention and for promising that all serialization issues are indeed thrown I'd suggest always throwing a `RuntimeException` when an `IOException` is hit.

Reply:
Basically, this is what I have in mind:
```java
  private interface JsonGeneratorRunnable {
    void run() throws IOException;
  }

  private static void generateOrThrow(final JsonGeneratorRunnable r) {
    try {
      r.run();
    } catch (final IOException e) {
      throw new RuntimeException(...);
    }
  }

  @Override
  public void serialize(final SerializedValue value, final JsonGenerator gen, final SerializerProvider provider)
  throws IOException
  {
    final var error = value.match(new SerializedValue.Visitor<Void>() {
      @Override
      public Void onNull() {
        generateOrThrow(gen::writeNull);
        return null;
      }
...
```

Reply:
> This means that callers can decide if they want to ignore the serialization error or not.

In this case no, because the visitor is only available to be used locally. This is just a workaround because the visitor interface doesn't allow throwing Exceptions. Outside of the visitor implementation, the returned exception is re-thrown. Since its an anonymous local implementation, no other code has the visibility to use that visitor.

Reply:
Oh! True.

I'd still avoid returning an `IOException` and instead use a wrapper method like `generateOrThrow()` to simplify things. I guess returning an exception instance isn't the worst thing but it feels like an anti-pattern.

Reply:
What I'm mostly getting at is that I don't think `serialize()` should `throws IOException`. At least I'm comfortable with it not throwing an `IOException` since Aerie's convention has been to throw an unchecked `RuntimeException`.

Reply:
So is the idea that `RuntimeException` doesn't need to be declared in the method signature in the `Visitor` interface?

Reply:
I guess I just don't understand why we'd need to convert from `IOException` to `RuntimeException`.

Reply:
Yup! I've updated the comment above, I mistakenly kept in the `final var error = ...`. https://github.com/NASA-AMMOS/aerie/pull/267#discussion_r930491839

Reply:
> I guess I just don't understand why we'd need to convert from `IOException` to `RuntimeException`.

Ahh I see! A `RuntimeException` is unchecked, whereas an `IOException` is checked. So when throwing a `RuntimeException` you don't need to add it to the `throws` list (nor do callers need to explicitly `catch` it).

Reply:
Ok lol this is too much :) I'm just gonna do it right and allow exceptions from inside the visitor

Reply:
Marking as resolved – I was thinking this was publicly called but it's just a Jackson-ism. Thanks for your patience Joel!